handbook/Best practices/TypeScript/Project-References.md

```diff
@@ -0,0 +1,166 @@
+    "baseUrl": "../../../",
+    "rootDir": "../../../",
+    "outDir": "../../../build/ts",
+    "resolveJsonModule": true,
+    "paths": {
+      "@shopify/package-b": ["packages/@shopify/package-b"],
+    }
+  },
+  "include": ["."],
+  "references": [{"path": "../package-b"}]
+}
+
+```
+
+### Running a build
+
+With the foundation in place, you can now run the TypeScript compiler
+
+```
+$ yarn run tsc -b config/typescript/project-references.tsconfig.json
+```
+
+> Protip: Use the ` --diagnostics` flag to get insight into what the compiler is doing. 
+
+It's very likely that you'll get a number of errors. The most common being that compiler can't find a module that your a project is referencing. The solution here would be to create a root level `tsconfig.json` for the module being depended on.Then add the reference to the `tsconfig.json` it in the project that fails to build. Keep in mind that you'll have the respect the following restrictions for project references:
+
+* The `include`/`files` compiler option must include all the input files that the project reference relies on.
+* Any project that is referenced must itself have a references array (which may be empty).
+* Specify the path mapping entries for module name to locations relative to the baseUrl. This is set in `#compilerOptions#paths`.
+
+### Next steps
+
+With a reasonable foundation in place, the rest of the codebase should be better equipped to start migrating over. Start by deciding which top-level folders you would like to migrate. Then gradually pick which sections you would like to convert over and create a `tsconfig.json` for that respective section. 
+
+```
+// app/sections/tsconfig.json
+
+{
+  "extends": "../../config/typescript/tsconfig.base.json",
+  "files": [],
+  "include": [],
+  "references": [
+    {"path": "./Home"}
+  ]
+}
+
+// app/sections/Home/tsconfig.json
+
+{
+  "extends": "../../../config/typescript/tsconfig.base.json",
+  "compilerOptions": {
+    "baseUrl": "../../../",
+    "rootDir": "../../../",
+    "outDir": "../../../build/ts",
+    "resolveJsonModule": true,
+    "paths": {
+      // List out all the project refeece path this section references
+    }
+  },
+  "include": ["."],
+  "references": [
+    // Include all the projec references this section relies on
+  ]
+}
+```
+
+
+## Tips
+
+* We've built a VSCode plugin ([TypeTrack](https://github.com/Shopify/typetrack)) for teams to leverage when migrating their codebase to Project References
+
+* Use relative imports when referencing code within it's own project reference. 
+
+Eg
+```
+// packages/@web-utilities/package-a/foo.ts
+
+// bad
+import {barThing} from '@web-utilities/package-a';
+
+// good
+import {barThing} from './bar-thing';
+```
+
+
+##  Caveats 
+
+Since depdendent projects make use of `.d.ts` files that are built from their dependencies you'll have to build a project after cloning it before VSCode can navigate the project in the editor without seeing spurious errors.
```

Comment:
```suggestion
Since dependent projects make use of `.d.ts` files that are built from their dependencies you'll have to build a project after cloning it before VSCode can navigate the project in the editor without seeing spurious errors.
```